### PR TITLE
Convert critical points by histogram worklets into filters

### PR DESCRIPTION
**Pull request by kmorel**

### CODE REVIEW

**Comment by kmorel** [comment]
@wangzhezhe, can you check to make sure that the `LoadEnsAndProcessByHistogram` example still works correctly? It would also be good to make sure that the execution time is comparable.

**Comment by wangzhezhe** [comment]
Sure, thanks a lot for the help! Let me look at the code in detail and test it

**Comment by kmorel** [comment]
FYI, I just pushed some changes to the CMake files. Hopefully that will allow it to be built by CUDA and HIP.

**Comment by wangzhezhe** [comment]
Sure, thanks for the information, there is a typo in the testing file here `ensemble2histogram.SetActiveField("emsembles",..)` After updating that, I got another error, maybe you know how to fix it quickly, I might need some time to understand the error message, thanks!

libc++abi: terminating due to uncaught exception of type vtkm::cont::ErrorBadType: Cast failed: vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<double>>, vtkm::cont::StorageTagRuntimeVec> --> vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>
4          0x187c590e0 start + 2360
3          0x1022bff58 main + 3192
2          0x1029ab718 vtkm::filter::uncertainty::CriticalPointsHistogram::DoExecute(vtkm::cont::DataSet const&) + 596
1          0x1029b7930 void vtkm::cont::UnknownArrayHandle::AsArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>>(vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>&) const + 336
0          0x1029b8140 void vtkm::cont::UnknownArrayHandle::BaseAsArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>(vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>&) const + 236

Loguru caught a signal: SIGABRT
Stack trace:
13         0x187c590e0 start + 2360
12         0x1022bff58 main + 3192
11         0x1029ab718 vtkm::filter::uncertainty::CriticalPointsHistogram::DoExecute(vtkm::cont::DataSet const&) + 596
10         0x1029b7930 void vtkm::cont::UnknownArrayHandle::AsArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>>(vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>&) const + 336
9          0x1029b8140 void vtkm::cont::UnknownArrayHandle::BaseAsArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>(vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>&) const + 236
8          0x108890cb0 vtkm::cont::throwFailedDynamicCast(std::string const&, std::string const&) + 232
7          0x187f93464 __cxxabiv1::exception_cleanup_func(_Unwind_Reason_Code, _Unwind_Exception*) + 0
6          0x187f93520 __cxa_get_exception_ptr + 0
5          0x187f90434 std::__terminate(void (*)()) + 16
4          0x187c2799c _objc_terminate() + 160
3          0x187f81110 demangling_terminate_handler() + 320
2          0x187f91070 __cxxabiv1::__aligned_malloc_with_fallback(unsigned long) + 0
1          0x187ee5a40 abort + 180
0          0x187fd9cc0 pthread_kill + 288
2024-03-19 14:03:23.766 (   0.129s) [main thread     ]                       :0     FATL| Signal: SIGABRT
Abort trap: 6

**Comment by kmorel** [comment]
I'll take a look and fix the problem.

**Comment by kmorel** [comment]
@wangzhezhe I just pushed some new code that should fix the problems. Several typos. Thanks for giving the information to run a test example. See if this works for you and does not cause problems with the runtime.

**Comment by wangzhezhe** [comment]
Sure, thanks for the help! I checked it just now, there is still this issue:

libc++abi: terminating due to uncaught exception of type vtkm::cont::ErrorBadType: Cast failed: vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<double>>, vtkm::cont::StorageTagRuntimeVec> --> vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>
4          0x187c590e0 start + 2360
3          0x1024cbf68 main + 3192
2          0x102bb90f0 vtkm::filter::uncertainty::CriticalPointsHistogram::DoExecute(vtkm::cont::DataSet const&) + 596
1          0x102bbccf4 void vtkm::cont::UnknownArrayHandle::AsArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>>(vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>&) const + 336
0          0x102bbdb74 void vtkm::cont::UnknownArrayHandle::BaseAsArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>(vtkm::cont::ArrayHandle<vtkm::VecFromPortal<vtkm::internal::ArrayPortalBasicWrite<float>>, vtkm::cont::StorageTagRuntimeVec>&) const + 236

File: ucvfilter/CriticalPointsHistogram.cxx [hunk 1]
```diff
@@ -0,0 +1,477 @@
...
+  // faster code (assuming you don't have to copy the array).
+  if (unknownHistogramDensity.CanConvert<vtkm::cont::ArrayHandleRuntimeVec<vtkm::Float32>>())
+  {
+    resolveType(
+      unknownHistogramDensity.AsArrayHandle<vtkm::cont::ArrayHandleRuntimeVec<vtkm::Float32>>());
+  }
+  else if (unknownHistogramDensity.CanConvert<vtkm::cont::ArrayHandleRuntimeVec<vtkm::Float64>>())
+  {
+    resolveType(
+      unknownHistogramDensity.AsArrayHandle<vtkm::cont::ArrayHandleRuntimeVec<vtkm::Float32>>());
```
Comment (wangzhezhe) [inline ucvfilter/CriticalPointsHistogram.cxx:461]: This usage is a little bit new to me,  should this be Float64, does that type cast error happens here?

Comment (kmorel) [reply ucvfilter/CriticalPointsHistogram.cxx:461]: Oops. That's a copy-paste error. That should both be `Float64`

Comment (kmorel) [reply ucvfilter/CriticalPointsHistogram.cxx]: This should be fixed now.

**Comment by wangzhezhe** [comment]
Sure, thanks a lot for the help, the code can work as expected now, thanks!

**Pull request merged**